compiler/rustc_query_system/src/dep_graph/graph.rs

```diff
@@ -556,7 +572,10 @@ impl<K: DepKind> DepGraph<K> {
             let mut edges = SmallVec::new();
             K::read_deps(|task_deps| match task_deps {
                 TaskDepsRef::Allow(deps) => edges.extend(deps.lock().reads.iter().copied()),
-                TaskDepsRef::Ignore => {} // During HIR lowering, we have no dependencies.
+                TaskDepsRef::EvalAlways => {
+                    edges.push(DepNodeIndex::FOREVER_RED_NODE);
+                }
+                TaskDepsRef::Ignore => {}
```

Comment:
Yeah, I've been wondering about that too. I think adding a dependency would be surprising. We are inside a `with_ignore()` after all. But ICEing might be useful?

One factor that might influence a decision here is that we seem to wrap everything inside of `TaskDepsRef::Ignore` by default:

https://github.com/rust-lang/rust/blob/a412564f8ca794e227305309da9da16c417ea9ce/compiler/rustc_middle/src/ty/context/tls.rs#L40-L49

If we changed that to `TaskDepsRef::Forbid` there would be less chance of things being silently ignored and as a result less harm in treating `TaskDepsRef::Ignore` as "yes, really turn off dependency tracking". We use `DepGraph::with_ignore()` in a few places where we really don't want any dep-tracking to happen, e.g. [when re-computing the value of a green node](https://github.com/rust-lang/rust/blob/a412564f8ca794e227305309da9da16c417ea9ce/compiler/rustc_query_system/src/query/plumbing.rs#L610).

But I haven't tried what the fallout of switching to `TaskDepsRef::Forbid` would be. It might need some cleanup.

Overall, I think `DepGraph::with_ignore()` should really reliable turn off dep-tracking -- but we should use it in as few places as possible. This PR would be an example of replacing it with something more targeted.